game/natsuki.rpy

```diff
@@ -1,376 +1,677 @@
+        def isWearingNecklace(reference_name):
+            """
+            Returns True if Natsuki is wearing the specified necklace, otherwise False.
+
+            IN: 
+                - reference_name - The necklace reference name to check if Natsuki is wearing
+
+            OUT:
+                - True if Natsuki is wearing the specified necklace, otherwise False
+            """
+            return Natsuki._outfit.necklace.reference_name == reference_name
+
+        # Start: Relationship functionality
+
+        @staticmethod
+        def calculatedAffinityGain(base=1, bypass=False):
+            """
+            Adds a calculated amount to affinity, based on the player's relationship with Natsuki and daily cap state.
+
+            NOTE:
+                If the player has not confessed to Natsuki via the talk_i_love_you topic, further affinity gain is not possible
+                until the confession has been made! This is to prevent players having Natsuki tell them she loves them unwarranted
+                by accidentally crossing the boundary into LOVE.
+
+            IN:
+                - base - The base amount to use for the calculation
+                - bypass - If the daily cap should be bypassed for things like one-time gifts, events, etc.
+            """
+            to_add = base * jn_affinity.get_relationship_length_multiplier()
+
+            if (
+                not persistent._jn_player_confession_accepted 
+                and (persistent.affinity + to_add) > (jn_affinity.AFF_THRESHOLD_LOVE -1)
+            ):
+                # Player cannot reach LOVE without having confessed to Natsuki successfully
+                persistent.affinity = jn_affinity.AFF_THRESHOLD_LOVE -1
+                jn_utils.log("Affinity blocked - CN!")
+                return
+
+            if bypass:
+                # Ignore the daily gain and just award the full affinity
+                persistent.affinity += to_add
+                jn_utils.log("Affinity+")
+
+            elif persistent.affinity_daily_gain > 0:
+                # Award the full affinity if any cap remains
+                persistent.affinity_daily_gain -= to_add
+                persistent.affinity += to_add
+
+                if persistent.affinity_daily_gain < 0:
+                    persistent.affinity_daily_gain = 0
+                
+                jn_utils.log("Affinity+")
+
+            else:
+                jn_utils.log("Daily affinity cap reached!")
+
+        @staticmethod
+        def calculatedAffinityLoss(base=1):
+            """
+            Subtracts a calculated amount from affinity, based on the player's relationship with Natsuki.
+
+            IN:
+                - base - The base amount to use for the calculation
+            """
+            persistent.affinity -= base * jn_affinity.get_relationship_length_multiplier()
+            jn_utils.log("Affinity-")
+
+        @staticmethod
+        def percentageAffinityGain(percentage_gain):
```

Comment:
Make sure this is very well balanced, it can break affinity gain quit heavily. But I'd remove such thing as % *gain*.